forge/test/operators/utils/plan.py

```diff
@@ -0,0 +1,465 @@
+                                            operator=input_operator,
+                                            input_source=input_source,
+                                            input_shape=input_shape,
+                                            dev_data_format=dev_data_format,
+                                            math_fidelity=math_fidelity,
+                                            pcc=test_collection.pcc,
+                                            kwargs=kwargs,
+                                        )
+
+                                        test_vector.failing_result = self._check_test_failing(test_vector)
+                                        yield test_vector
+
+
+@dataclass
+class TestParamsFilter:
+    '''
+    Dataclass for specifying test parameters filter
+
+    Args:
+        allow: Allow function
+        indices: Indices to filter
+        reversed: Reverse the order
+        log: Log the parameters
+    '''
+    allow: Optional[Callable[[TestVector], bool]] = lambda test_vector: True
+    indices: Optional[Union[int, Tuple[int, int], List[int]]] = None
+    reversed: bool = False
+    log: bool = False
+
+
+class TestPlanUtils:
+    '''
+    Utility functions for test vectors
+    '''
+
+    @classmethod
+    def _check_kwargs_matching_list (
+        cls,
+        kwargs_def_list: List[OperatorParameterTypes.Kwargs],
+        kwargs: OperatorParameterTypes.Kwargs,
+    ):
+
+        for kwargs_def in kwargs_def_list:
+            if cls._check_kwargs_matching(kwargs_def, kwargs):
+                # One of the rules matched so collection is valid
+                return True
+
+        # No rule matched
+        return False
+
+    @classmethod
+    def _check_kwargs_matching (
+        cls,
+        kwargs_def: OperatorParameterTypes.Kwargs,
+        kwargs: OperatorParameterTypes.Kwargs,
+    ):
+        for(kwarg_name, kwarg_def_value) in kwargs_def.items():
+            if kwarg_name not in kwargs:
+                # logger.warning(f"Rule defined for kwarg {kwarg_name} not present in test vector")
+                return False
+            kwarg_val = kwargs[kwarg_name]
+            if isinstance(kwarg_def_value, tuple):
+                if not (kwarg_def_value[0] <= kwarg_val <= kwarg_def_value[1]):
+                    # logger.warning(f"Kwarg value out of range: {kwarg_name}={kwarg_val} not in {kwarg_def_value}")
+                    return False
+            else:
+                if kwarg_def_value != kwarg_val and not (kwarg_def_value is None and kwarg_val is None):
+                    # logger.warning(f"Kwarg value mismatch: {kwarg_name}={kwarg_val} != {kwarg_def_value}")
+                    return False
+
+        # All kwargs checked and no mismatch found
+        return True
+
+    @classmethod
+    def test_vector_in_collection(cls, test_vector: TestVector, test_collection: TestCollection, check_failing_reason: bool = False) -> bool:
+        return \
```

Comment:
Can we simplify this and make it more readable? Just splitting these conditions in named functions and calling them with conditions. E.g.
```py
def test_vector_in_collection(...):
    def match(attribute, vector_value):
        return attribute is None or vector_value in attribute

    def match_failing_reason(reason_attr, vector_reason):
        return not check_failing_reason or (reason_attr is None or (vector_reason is not None and reason_attr == vector_reason))

    return (
        match(test_collection.operators, test_vector.operator) and
        match(test_collection.input_sources, test_vector.input_source) and
        ...
        match_failing_reason(test_collection.failing_reason, test_vector.failing_result.failing_reason) and
        match_failing_reason(test_collection.skip_reason, test_vector.failing_result.skip_reason)
        ...
    )
```

Reply:
I like this idea very much. I'm only concerned about nesting match functions in python since they are called many times from surrounding loops.

Reply:
Refactor completed